packages/postgres-database/tests/test_utils_folders_v2.py

```diff
@@ -0,0 +1,1856 @@
+    )
+    editor_a_folder_id = await folder_create(
+        connection, "editor_a_folder", editor_a_gid, parent=d_folder_id
+    )
+    await _assert_folder_entires(connection, folder_count=5, access_rights_count=9)
+    # share existing folder in hierarchy with a new user
+    await folder_share_or_update_permissions(
+        connection,
+        d_folder_id,
+        sharing_gid=owner_a_gid,
+        recipient_gid=editor_b_gid,
+        recipient_role=FolderAccessRole.EDITOR,
+    )
+    await _assert_folder_entires(connection, folder_count=5, access_rights_count=10)
+
+    # FOLDER STRUCTURE {`fodler_name`(`owner_gid`)[`shared_with_gid`, ...]}
+    #  `root_folder`(`owner_a`)[`owner_b`,`owner_c`,`owner_d`,`editor_a`]
+    #   - `b_folder`(`owner_b`)
+    #   - `c_folder`(`owner_c`):
+    #       - `d_folder`(`owner_d`)[`editor_b`]:
+    #           - `editor_a_folder`(`editor_a`)
+
+    # check top most parent resolution
+    async def _assert_reloves_to(
+        *,
+        target_folder_id: _FolderID,
+        gid: _GroupID,
+        permissions: _FolderPermissions,
+        expected_folder_id: _FolderID,
+        expected_gids: set[_FolderID],
+    ) -> None:
+        resolved_parent = await _get_top_most_access_rights_entry(
+            connection,
+            target_folder_id,
+            gid,
+            permissions=permissions,
+            # NOTE: this is the more restricitve case
+            # and we test against exact user roles,
+            # the APIs use only a subset of the permissions ususally set to True
+            enforece_all_permissions=True,
+        )
+        assert resolved_parent
+        assert resolved_parent["folder_id"] == expected_folder_id
+        assert resolved_parent["gid"] in expected_gids
+
+    await _assert_reloves_to(
+        target_folder_id=root_folder_id,
+        gid=owner_a_gid,
+        permissions=OWNER_PERMISSIONS,
+        expected_folder_id=root_folder_id,
+        expected_gids={owner_a_gid},
+    )
+    await _assert_reloves_to(
+        target_folder_id=b_folder_id,
+        gid=owner_b_gid,
+        permissions=OWNER_PERMISSIONS,
+        expected_folder_id=root_folder_id,
+        expected_gids={owner_b_gid},
+    )
+    await _assert_reloves_to(
+        target_folder_id=c_folder_id,
+        gid=owner_c_gid,
+        permissions=OWNER_PERMISSIONS,
+        expected_folder_id=root_folder_id,
+        expected_gids={owner_c_gid},
+    )
+    await _assert_reloves_to(
+        target_folder_id=d_folder_id,
+        gid=owner_d_gid,
+        permissions=OWNER_PERMISSIONS,
+        expected_folder_id=root_folder_id,
+        expected_gids={owner_d_gid},
+    )
+    await _assert_reloves_to(
+        target_folder_id=editor_a_folder_id,
+        gid=editor_a_gid,
+        permissions=EDITOR_PERMISSIONS,
+        expected_folder_id=root_folder_id,
+        expected_gids={editor_a_gid},
+    )
+    await _assert_reloves_to(
```

Comment:
reloves??? :heart:

Reply:
such a lovely typo and I have to change it.